### DISCUSSION TITLE
"No available authenticator recognized any of the allowed credentials" in iOS, even in example app

### DISCUSSION THREAD

**Discussion by cannikin** [MasterKale/SimpleWebAuthn#211] [Troubleshooting]
Got my implementation working beautifully in Chrome and Safari on the desktop, no problems.

Moving to iOS (15.5), registration works, but when I go to authenticate I get `No available authenticator recognized any of the allowed credentials` my `catch` block around the `startAuthentication()` function. I thought maybe something about my specific implementation was messed up so I tried in the example app and I see the same thing:

<img src="https://user-images.githubusercontent.com/300/172710886-8cffc18b-863c-476d-8c8e-10ddbaa812c2.jpg" width="300">

Anyone else seen this?

**Comment by lmarschall** [comment]
What kind of authenticators are you using on desktop and on iOS? Maybe it‘s got something to do with the allowCredentials flag on the server side?

**Comment by cannikin** [comment]
On the desktop I have the `authenticatorAttachment` set to `platform` which means TouchID on my Macbook. Doing the same thing for my iPhone 13 means FaceID.

**Comment by MasterKale** [comment]
I think I know what's going on: Safari typically allows websites a single "freebie" indirect execution of WebAuthn. If there was any kind of asynchronicity before your call to one of the `navigator.credentials` then Safari would fail the API call right away with an inscrutable error message.

Fortunately sometime back in February Apple eased up on that, and now websites were more free to make API requests and perform other actions before ultimately calling WebAuthn, but only once for a given page load. After that you'd potentially have to reload the page to get WebAuthn to reliably trigger again.

I tested my example site on iOS too and no matter which button I clicked first, the first button would act as expected, while the second button would always error out in the same way you described. This is making me believe that you're describing that Safari limitation.

**Comment by cannikin** [comment]
Hmm, is there any workaround for that, then? This is in React, so there's potentially an infinite number of events before any WebAuthn invocation! `startAuthentication()` itself is async and wraps the calls to `navigator`, what are we supposed to do if you need to call `navigator` directly in an `onClick`?

**Reply by MasterKale** [reply]
"Indirect" means the call isn't made in the callstack of an event handler. For example, eventually calling `navigator.credentials.get()` from within a `setTimeout()` is going to run into the restriction more than, say, an async method defined on a button click handler that calls that method through something like SimpleWebAuthn's `startAuthentication()`.

I haven't sat down and tried to find the boundaries of Safari's restrictions around this. And who knows how it'll change with conditional UI in Safari starting in macOS Ventura. I'll try to make some time to more purposefully explore what limitations SPA's have in invoking WebAuthn. Any observations you note down will be helpful too in the name of confirming our findings.

**Reply by cannikin** [reply]
Hmm, I'm still don't have a good mental picture of "indirect." Right now my flow is:

User clicks a button labeled "Open Authenticator" with a React `onClick` prop:

```jsx
<button onClick={authenticate}>Open Authenticator</button>
```

And `authenticate()` does the following:

1. `await fetch()` to my server to retrieve the options object
2. `await response.json()`
3. `await startAuthentication(options)`

Is that "too many" awaits and now it's considered indirect? I could get the options as soon as the page loads to remove that from the call stack when clicking the button...maybe that would help?

**Reply by MasterKale** [reply]
I wish I could offer more guidance here, but I'm no WebKit engineer so I can't look at the code to confirm and Apple definitely hasn't explicitly laid out the rules anywhere. The best I can offer is trial-and-error to try and determine what constitutes "indirect" execution...



> And `authenticate()` does the following:
> 
> 1. `await fetch()` to my server to retrieve the options object
> 2. `await response.json()`
> 3. `await startAuthentication(options)`

Is this not working _at all_ for you in Safari? Or does the first click work and then subsequent attempts fail? Off hand I'd say nothing you describe here should prevent WebAuthn from working. SPA's are unique of course in that the page doesn't ever reload. 

All that said it's also a possibility that this could be a legitimate issue with WebKit, but it'd require more testing to feel any confidence in that idea 🤷‍♂️

**Reply by cannikin** [reply]
The weird thing is that it only happens on *authenticate* and not on *register*...you'd think the behavior would be the same?

If I go to my app, go to login page, login with username/password, ask if I want to enable webauthn, click a button, scan FaceID and register, everything works! That's a ton of clicks and navigation before getting to a `navigator.credentials` call, right?

Authenticate is a little trickier. Once I'm logged in, if I then log out, click a link to return to the login page, click my "Open Authenticator" button, I get the "No available authenticator" error:

https://user-images.githubusercontent.com/300/173154065-ae08e99b-beaa-48c0-97b0-0b818f315ef7.mov

However, if I refresh the page, then click "Open Authenticator" it works. Similar to the behavior you mentioned, where it seems like it has to be the first button you click (but again, strange that it doesn't seem to be the case for registering):

https://user-images.githubusercontent.com/300/173154081-1a908512-74ff-4cc3-a6d4-462e08b8fce0.MP4

I've noticed that when the "Do you want to let this site use FaceID?" prompt comes up, if you click cancel, you get the same "No available authenticator" error message. Maybe that's how Safari treats that second/third/fourth click—as if you choose to decline using FaceID.

**Reply by MasterKale** [reply]
So here's some interesting console output I noticed today with my example site:

![Screen Shot 2022-06-13 at 1 29 04 PM](https://user-images.githubusercontent.com/5166470/173440081-420b53ff-0f14-4ef4-9906-d3a38c8b8cdf.png)

No matter which button I clicked first, the second button would throw that "User gesture is not detected" error. So I guess this confirms that there's something about the event handler that's just indirect enough to trip WebKit's restrictions on indirect execution 🤔

**Reply by cannikin** [reply]
I noticed that as well...I saw something on Stack Overflow about using an `onTouchStart` instead of `onClick`? I didn't actually get a chance to implement that though. I wasn't sure if I was going to have to go down some rabbit hole of checking for user agent or some other method to determine whether to use `onTouchStart` or `onClick` depending on the browser/platform. :(

**Reply by MasterKale** [reply]
That doesn't sound like a solution at all, according to MDN the `onTouchStart` event is only supported on mobile 😂 😭 

https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/ontouchstart#browser_compatibility

<img width="785" alt="Screen Shot 2022-06-14 at 9 10 07 PM" src="https://user-images.githubusercontent.com/5166470/173735063-9037ba03-32e7-4204-9212-497d82ca1b4c.png">

**Reply by cannikin** [reply]
Here's another weird behavior...so when I get the "No available authenticator..." message, if I reload the page then the call to open the authenticator works fine (even when that call is not user-initiated, I have it fire off as soon as the page loads). I figured I could just catch that "No authenticator available" error and then `location.reload()`, easy peasy! But nope, I get in an infinite reload loop, it keeps triggering the error, which reloads the page, which triggers the error...

It seems like the reload *itself* needs to be user initiated, even if the call to start the auth process isn't?? I even tried making my fallback button that says "Open Authenticator" have an `onClick={() => window.location.reload()}` and it does the same thing, throws the error again. It's like the user **has** to click the "reload" button in the browser itself. This is the same behavior in iOS Chrome in addition to Safari.

What the hell is going on?

**Reply by cannikin** [reply]
More testing...if I open my login page with an `<a>` tag (causing a real page load, rather than letting React just move pages in and out of the history) then I get FaceID to launch! However, If I then log out and click the <a> link again to go back to login, now it won't work (even though that <a> click is causing a real page load again!). 🫤

**Reply by MasterKale** [reply]
Something something "user security" I suppose. 

> It seems like the reload itself needs to be user initiated, even if the call to start the auth process isn't?? I even tried making my fallback button that says "Open Authenticator" have an onClick={() => window.location.reload()} and it does the same thing, throws the error again. It's like the user has to click the "reload" button in the browser itself. This is the same behavior in iOS Chrome in addition to Safari.

I can see how they came to this conclusion, a malicious site could endlessly spam the user to complete WebAuthn if it somehow got access to the page (bad lib import, etc...) and could programmatically reload the page to try again if the user cancelled out.

This still makes it difficult to intuit what constitutes a sufficient "user gesture"...

**Comment by MasterKale** [comment]
I just converted this issue into a discussion since it's been confirmed to be a WebKit restriction on the use of WebAuthn.

**Comment by fromtexas** [comment]
Got the same problem. I changed fetch to XMLHttpRequest and the problem disappeared 🪄

**Reply by cannikin** [reply]
Oh snap really?? I just assumed fetch was a nicer abstraction over XMLHttpRequest...what'd you do about headers, credentials, etc?

**Reply by fromtexas** [reply]
XMLHttpRequest has boolean property [withCredentials](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials) for credentials and also method [XMLHttpRequest.setRequestHeader()](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/setRequestHeader) for headers. But, I guess, [axios](https://github.com/axios/axios) would be better option than using raw XMLHttpRequest

**Reply by cannikin** [reply]
Oh snap just re-write my implementation to XMLHttpRequest and sure enough, iOS Safari is happy! Thanks so much for following up and posting in this thread! :)

**Reply by MasterKale** [reply]
I...I'm glad you found a solution but also horrified that you found _this_ solution. So use of `fetch()` enforces the indirect user gesture requirement, but `XMLHttpRequest` doesn't?

I dunno that I'd build anything atop this knowledge, it sounds more like a bug than a feature 🤔

**Reply by cannikin** [reply]
😈
